Hordes 2.0 again!

kevingranade opened the pull request:
#### Summary
Features "Implement hordes for longer ranged interaction with monsters."

#### Purpose of change
Finally, a follow-up to the ancient #14240 
In short, this feature makes monsters that exist within TBD distance of the reality bubble able to navigate fluently around, more or less as if they were in the reality bubble and appear at the edge of said bubble when they intersect with it.
This should make the world *feel* larger as certain actions such as making lots of noise (gunfire, driving vehicles, running generators) should now attract monsters from greater distances.
Additionally, but likely not in the first version of the feature, this gives us the opportunity to give monsters behaviors of their own independent of the player's presence.  For example a common and effective strategy is leading hordes of zombies away from a building you want to raid.  It should be feasible to add a behavior where zombies "prefer" cities and will trickle back into habitable areas over time once something has scattered them.
This will also make monsters more persistent and harder to run away from since they can continue to move even after you are 60+ tiles away from them, when currently they freeze in place instantly. This if course will come with rebalancing of player and monster move speeds and endurance.

#### Describe the solution
I've established a bitmap of values that mark whether a given map space is passable or not that is stored at the overmap level.
I've also adjusted the logic for storing monsters at overmap scale. Once a monster leaves the reality bubble, either because of their own movement or player movement, (usually the latter), they are removed from the local data structures and moved to this overmap-scale structure.
Periodically, nearby monsters stored in this structure are refreshed to move toward their current destination. If this movement moves them into the reality bubble, they are spawned and can proceed from there as usual.

#### Describe alternatives you've considered
I can't really think of any direct alternatives, but what has changed from previous iterations of this is I'm focusing on just getting something working and iterating instead of trying to plan out every detail.
Both sets of new data structures are incomplete and extremely inefficient in both space and execution time.
The code to generate the "passable tiles" data is a shoddy placeholder at best (I will be addressing this in this PR though)
Something to be aware of relative to the scope of this PR in particular is that it is unlikely to dramatically change game balance.  What I've seen in testing is that monsters aren't that much more persistent than they already are now, because by the time the player outruns them and leaves them behind, they've nearly forgotten about the player already.  For this PR I want to get this feature WORKING and delay balance changes such as monsters being more persistent at chasing you for later. Likewise, attracting them via loud noise requires VERY loud noises, and hardly anything is actually that loud since there is no reason to set them that high other than deafening the player because any noise over 60 would not attract any additional monsters before this PR.

#### Testing / feature outline
- [x] Spawn a monster outside the reality bubble.
- [x] Give the monster a goal in the middle of the reality bubble.
- [x] Advance time until the monster reaches the bubble and spawns.
- [x] Continue advancing time until the monster reaches the center of the bubble.
- [x] As above but alert the monster via a loud sound instead of manually setting its destination.
- [x] Add some basic per-OMT placeholder bit patterns to reflect "flat road", "underbrush", "cluttered forest", "omt-filling building" and perhaps a few more.
- [x] Add logic to store the actual state of any OMT modified (building, breaking, etc) while loaded in the reality bubble as an actual bitmap and store it in the overmap when the omt is unloaded.
- [x] Serialize the obstacle bitmap, probably as a base64 string per OMT?
- [x] Shrink the data structure storing monsters, some napkin math says the current data structure will result in a MB+ of monster data per overmap, which is "a bit much".
- [x] Add some migration code to move "monster group" entries to the new monster_map data structure.
- [x] Do some performance testing to see what kind of horde update radius we can handle. Probably the worst case for this is "create an absolutely enormous sound that alerts all the loaded hordes and then go to sleep" to minimize competing sources of slowdown.
- [x] Evaluate throttled horde updating.  ~~What I'm thinking is we draw radial segments extending from the player location and process whole submaps worth of hordes within those segments per turn. Then each turn we shift to a new segment.  In an area immediately around the reality bubble we would process all submaps more frequently so horde zombies arrive at accurate times. The radial update strategy optimizes for monsters moving toward the player, which is the most likely pattern.~~ This general concept is valid, but actual spokes triggers far more updates, so it's not generally applicable.  Possibly we can chunk up the map so that we do radial line segment updates of limited length to minimize over-updating. Upon more thought, probably the simplest batching logic we can do is something like "Walk all the OMTs and ignore any that don't pass a check that's something like if( ( location.x + location.y + turn ) % distance_to_reality_bubble_in_ms() ) This is less concerned with attempting to insure that we absolutely minimize how many monster_map and map_data entries we load and just prioritizing batching movement more and more aggressively the further the horde is from the player.
- [x] Adjust the "see hordes at distance" code to use monster_map instead of monster groups.
- [ ] Test and fix up scenarios around monsters transiting through an OMT when it is shifted into and out of.
- [ ] Test the specific scenario where the player is fleeing noisily from a monster while matching it's speed, the monster should be able to pursue indefinitely if not out-distanced.
- [x] Add debug visualization for overmap-resident monsters (probably already exists but is pointed at the wrong data structure I think).
- [x] Debug monster spawning and removal features for overmap level.
- [ ] Add more nuanced move costs instead of just pasable/impassable, tune up movement to better match how monsters act while in the bubble. Update to this, I was originally assuming I would want to put per-tile move costs in this data structure, but upon further reflection that doesn't seem at all necessary, an average move cost per tile should be fine. This is good because this was going to be one of the most expensive pieces of data this new feature would generate.
- [ ] Consider some basic pathing so monsters don't get stuck constantly, if nothing else just detect being stuck and wander for a bit before resuming the chosen path.
- [ ] evaluate some performance alternatives such as setting a path periodically and only revisiting the monster after some progress is made instead of refreshing every turn.
- [ ] as above, but also evaluate things like refreshing nearby monsters more frequently.
- [x] make sure monsters that spawn out of hordes have evolution applied.

kevingranade committed:
Initial version of cache of map data at overmap scale

kevingranade committed:
Add a method to spawn a monster directly into the overmap monster_map

kevingranade committed:
Overhaul overmap::move_hordes to use new system.

kevingranade committed:
Add overmap passability cache clearing to map_helper::clear_map()

kevingranade committed:
Add tests for monsters ability to navigate across the reality bubble and passing from the overmap to the reality bubble.

GuardianDll commented:
> Also note that sound propagation range should be shortened in a similar proportion to the firing range shortening currently used.

Our noise propagation is already extremely short, 556 would generate about 120 noise, which is only like 120 meters, when in reality 556 would be heard for at least a few kilometers around

Profugo-Barbatus commented:
Exciting! If I'm understanding the OMT passage right, the goal would be a player could build a wall across a line of OMT's, and a horde on the far side would have to navigate around it. And if the player approached the wall from the far side of the horde, the entering undead would only appear on the opposite side of the wall. If correct, being able to build large scale and not have zombies teleport into seemingly inaccessible areas will be great for faction camps and other large player constructions/renovations. 

Either way, very much looking forward to this. Existing horde behavior is a little braindead, sometimes it feels like I'm chasing them and not the other way around. Improved navigation sounds like a great first step towards enabling them to be more aggressive.

kevingranade committed:
Move omt map sumary to own file, add placeholder store

kevingranade commented:
Slow progress, have omt data placeholder v1 implemented, next up is attaching placeholders to OMTs by populating that new oter_type_t::default_map_data member and using that to populate the main data structure.

kevingranade commented:
Quite a big obstacle, I was hoping to use oter_type_t::travel_cost_type to populate default_map_data, but there are ~5,000 oter_type_t instances missing this field.
For now I'm punting, but this is going to have to be resolved somehow, which might mean many hundreds to thousands of json entities getting edited.

kevingranade committed:
Sins against memory management

Also implements using both placeholders and dynamic map_data_summary objects in overmap.

XingXingMonkeyGrandma commented:
> Quite a big obstacle, I was hoping to use oter_type_t::travel_cost_type to populate default_map_data, but there are ~5,000 oter_type_t instances missing this field. For now I'm punting, but this is going to have to be resolved somehow, which might mean many hundreds to thousands of json entities getting edited.

I would actually find a tedious project like this kind of fun. Happy to help if it would be welcome

akrieger commented on src/overmap.cpp:
```diff
@@ -3170,6 +3203,11 @@ oter_id overmap::get_default_terrain( int z ) const
 }
 
+// underlying bitset default constructs to all 0.
+static std::shared_ptr<map_data_summary> impassable_omt( new map_data_summary() );
+static std::shared_ptr<map_data_summary> passable_omt( new map_data_summary(
+            ~impassable_omt->passable ) );
```
This is what the 'nonempty 0 alloc shared_ptr trick' is best for, because below you make a ton of increfs on the statics.

```suggestion
static map_data_summary impassable_omt{};
static map_data_summary passable_omt{ ~impassable_omt->passable ) };
```

akrieger commented on src/overmap.cpp:
```diff
@@ -3178,6 +3216,8 @@ void overmap::init_layers()
         l.visible.fill( om_vision_level::unseen );
         l.explored.fill( false );
+        // Verify this isn't copying!
+        l.map_cache.fill( passable_omt );
```
```suggestion
        l.map_cache.fill( std::shared_ptr<map_data_summary>{ nullptr, &passable_omt } );
```

akrieger replied:
Yeah a dummy variable like the other place, whups.

akrieger commented on src/overmap.cpp:
```diff
@@ -3604,6 +3646,100 @@ std::vector<point_abs_omt> overmap::find_extras( const int z, const std::string
+    }
+    if( ptr->placeholder ) {
+        // Copy the placeholder data.
+        ptr = std::make_shared<map_data_summary>( ptr.get() );
```
If you only want to copy the placeholder then use the constructor you already had for that. But I think you mean to make a deep copy here?
```suggestion
        ptr = std::make_shared<map_data_summary>( ptr->placeholder );
```

kevingranade replied:
No just the data,  and yea that's so much cleaner.

akrieger replied:
Edited the code suggestion to be, um, correct.

akrieger reviewed (commented):
I haven't looked at all at overmap_map_data_cache yet.

akrieger commented on src/overmap_map_data_cache.h:
```diff
@@ -0,0 +1,42 @@
+{
+void load( const JsonObject &jo, const std::string &src );
+void reset();
+void finalize();
```
```suggestion
```

kevingranade commented on src/overmap_map_data_cache.h:
```diff
@@ -0,0 +1,42 @@
+struct map_data_summary {
+    map_data_summary() = default;
+    map_data_summary( std::bitset<24 * 24> new_passable ): passable( new_passable ) {}
+    map_data_summary( map_data_summary *new_summary ): passable( new_summary->passable ) {}
```
No that's intentional, this is to implement copy-on-write of the bitmap but leave everything else with default values.

kevingranade committed:
Add map::on_unload() to handle OMTs leaving the map

kevingranade commented:
Next up is actually grab live map data and stuff the bitmap into the overmap when we detect that the player has probably edited the map.

Plan for this is add a bool player_adjusted_map field to submap.
In map::ter_set() an map::fur_set() (maybe others?) we set this bool by default when we edit a submap.
This behavior is suppressed by a variable in map::mapgen_in_progress that is set and cleared by mapgen functions such as map::generate()

To test we should be able to just run mapgen a bunch and then assert that no submaps have the flag set?  That should flush everything out.

Later we can consider some kind of heuristic for whether the change is "meaningful" instead of any time a submap is changed outside of mapgen.

kevingranade committed:
Replace placeholder map_data_summaries with real data as needed

Marks submaps when they are edited outside of mapgen

Upon unloading marked submaps from the reality bubble,

write their data to a corresponding map_data_summary.

kevingranade commented:
Re: Test and fix up scenarios around monsters transiting through an OMT when it is shifted into and out of.
This does break, though not for any exotic reason.  If you walk toward a horde you can trigger debugmsg( "there's already a monster at %s", critter.pos_abs().to_string_writable() ); in creature_tracker::add()

Not sure why it's happening, would prefer not to have a goofy workaround like delaying spawn or shifting monster location.

kevingranade committed:
Change structure of overmap::monster map to be a std::unordered_map<std:map<monster>>

This retains the bulk lookup feature while enforcing one monster per tile

kevingranade commented:
Took a lot of changes, but replaced `std::unordered_multimap<tripoint_om_sm, monster> monster_map`
with
`std::unordered_map<tripoint_om_sm, std::map<tripoint_abs_ms, monster> monster_map`
and augmented overmap::passable() to check if there is a monster present in a given tile, which means horde members will never enter the same tile.  This is how it should work anyway, because for example if navigating through a city using gaps between buildings, a sufficiently large horde will have to squeeze between buildings and slow down when chasing you.

This also means that every time a horde member moves, it is removed from monster_map and re-inserted. This is done as a *splice* though and avoids allocation.

Doing this also surfaced another issue which is that having entries move around like this runs the risk of a monster being processed multiple times per turn. Avoided this by setting a "monster last processed on X turn" value, and if a monster is encountered that already has this updated to the current turn they are skipped.  This will be useful in the future when we want to process hordes in batches and "catch them up".

Finally, tested that running from a horde works, both in the situation where you get ahead then make noise it calls them back onto the map, and in the situation where you walk back toward them and they load back onto the map without error.

There is some kind of issue where they can "pop" into an already vissible area instead of being loaded along with the submap, not sure what's happening there yet.

PatrikLundell commented:
We've had issues in the past where creatures spawn where they shouldn't, and haven't been able to track it down. It's tricky, because when it happens it's already too late, as you need a save just before it happens to be able to examine it, and it doesn't happen often enough to really make it possible to catch it by saving frequently.

If you can get it to happen reliably we may get lucky and catch this bugger (or at least one instance of it).

A few possible causes:
- Using the wrong map with the map relative coordinates. This shouldn't happen anymore, but there may well be bugs there.
- Using PC relative coordinates rather than map relative ones, or vice versa.
- Using the map relative coordinates of a location before a map is shifted on the map after it has been shifted. That should only shift locations by one submap, though.

A couple of weird cases that have happened to me:
- Mi-go prisoners raining out of the sky at the edge of the reality bubble, despite there being any tower anywhere near (they presumably fell from the upper floor when they spawned in open air rather than in a tower).
- A bunch of web spiders that suddenly spawned in front of my car when traveling along a route traveled quite a few times.
- A pair of web spiders that decided to rapidly travel to and settle on the road elsewhere. Those may have been lured there for some unknown reason from a spider web area fitting within the reality bubble, though.

db48x commented:
> It's tricky, because when it happens it's already too late, as you need a save just before it happens to be able to examine it…

If I might interject to disagree with you slightly, it should be possible to use a record & replay, or time travel, debugger such as [rr](https://rr-project.org/) to catch the cause of this problem.

A debugger of this type allows you to record the program as it runs, then play it back later in a debugger. You would be able to break into the debugger once you spot a monster spawning in an incorrect position, set a breakpoint on the function that spawns it, then run the program backwards to the exact moment when the monster is actually placed onto the map. You can then inspect the state of the program to discover the coordinates in use. You can trace the computation of those coordinates backwards to discover how they were determined.

There is thus no need to have a save from seconds before the bug occurs in order to observe the problem in a debugger.

> …and it doesn't happen often enough to really make it possible to catch it by saving frequently.

This is still a difficulty. It would be necessary to record the game every time you play it, waiting and hoping to spot the bug as it happens. Once you do spot the bug, however, you can preserve the recording and replay it as many times as is necessary to track down the bug. It is even possible to pack up the recording to be sent to others so that they can debug the same problem as you.

Keep in mind then that different debuggers implement this feature in different ways. GDB has a built–in recorder, but it is between one hundred thousand and a million times slower than real time. This makes it completely unsuitable for this purpose. The one I recommend, [rr](https://rr-project.org/), tends to slow down the program only by about 40%; most of the time this would hardly be noticed as the game is waiting for player input in any case.

Feel free to ask me questions about rr; I’d be happy to assist.

PatrikLundell commented:
I have to admit I wasn't aware of the ability to record and replay progress in this way. Still the low frequency is a real issue. I've seen it once per month, if at that, and 40% is still quite a significant slowdown for a game that's subject to very significant slowdowns at times.

db48x commented:
Suit yourself.

kevingranade committed:
Add test for triggering horde spawning with player movement

kevingranade committed:
Add horde_entity wrapper that allows never-spawned monsters to have tiny entries instead of an entire monster object.

Blueflowerss commented:
there's an issue (or was? i haven't been keeping track) with sea bases and hordes. Where hordes would travel over the sea map tiles to your location, spawn into the reality bubble, sink to the lowest z-level and then just wander there. In one of my playthroughs this resulted in a few dozen zombies just chilling at the bottom of the lake under my base, slowing down the game immensely.
just wanted to mention this somewhere.

kevingranade committed:
When loading mongroups shift hordes to monster_map
Also place monsters in monster_map instead of in the mon group map.

kevingranade commented:
Adjusted monster group loading code to spawn monster groups with horde == true into monster map instead of zg.
Next up is shift horde visualization over to monster_map, then use that to take a look at how many monsters are getting placed on a legacy overmap and a new overmap.

kevingranade committed:
WIP Move horde debug functions to target monster_map

kevingranade committed:
Add tests for overmap map data cache

kevingranade committed:
Shift hordes to be based on a new monster group with no panicked people in it.

Also skip over DORMANT monsters when moving hordes

kevingranade commented:
I was noticing that if I made a bunch of noise then waited without moving, the monsters weren't making it to me, debugging showed they were hanging out on the edge of the map and preventing other horde members from spawning.
After a few tests I noticed that they would start out coming right for me, then *turn around* and head the other way, when I checked the list of spawned monsters I saw that panicked people were spawning with the horde and distracting them when they spawned onto the map.
I made a new monster group ZOMBIE_HORDE that doesn't have panicked people in it, and added code to swap out legacy monster groups that are from the ZOMBIE group to ZOMBIE_HORDE at the point where they are migrated to the new horde data structure.

WHAT IS LEFT is a bit less than what is implied by the checklist, perhaps a LOT less.
I *definitely* need to serialize player modified map data.
I need to take a last crack at characterizing overhead from the new hordes code, it's possible it needs little to no special handling for performance, and can operate across the current horde update zone which is... uh what is it actually... the entire current overmap plus any other overmaps with a border within 22 submaps / 134 tiles of the player?  I was anticipating it being unplayably bad, but it's definitely not.
Definitely need to add horde sighting at a distance, it's already done for debug stuff but I need to switch over the player-oriented version too.
That's it, those three things.

kevingranade commented:
Tested the encoding and decoding of bitsets to base64 strings, it was horrifically broken of course, but works now.

Still needs a (manual AFAIK) full round trip test that player-built structures get stored and restored and work to block horde motion.

kevingranade committed:
Add test for persisting player-made changes to map in overmap data

Also fix an index scaling bug that was breaking it

kevingranade commented:
Manual round trip test and subsequent fixes are a mixed bag.
Good news, I was able to complete the scenario where I place a huge wall then make a ton of noise, hordes this attracts are then blocked by the wall even though neither are in the reality bubble.
More good news, when I examined the save file it had 69MB of map placeholder data that it wasn't supposed to be writing out, but I tracked that down to bad placeholder data initialization and it's not doing it any more.

Bad news, examining the save file and what it's writing is completely wrong and I don't know why.

This is a base64 encoded bitset, '-' is the "all bits set" character, so these OMTs are registering as almost totally passable?
The two substrings 38AAAAA indicate a run of approximately 30 unpassable tiles, which is *almost* right, but the walls I made were about 121 tiles long, not 30, and there should be a big vertically oriented wall too, which would show up as a bunch of scattered 'B' characters at regular intervals.
```
"map_data":[[117,159,0],"#------------------------------------------------------------------------------------------------",[118,159,0],"#------------------------------------------------------------------------------------------------",[119,159,0],"#------------------------------------------------------------------------------------------------",[120,159,0],"#------------------------------------------------------------------------------------------------",[121,159,0],"#------------------------------------------------------------------------------------------------",[122,159,0],"#------------------------------------------------------------------------------------------------",[115,160,0],"#-------------------------38AAAAA---------------------------------------------------------38AAAAA",[117,160,0],"#------------------------------------------------------------------------------------------------",[118,160,0],"#------------------------------------------------------------------------------------------------",[119,160,0],"#------------------------------------------------------------------------------------------------",[120,160,0],"#------------------------------------------------------------------------------------------------",[121,160,0],"#------------------------------------------------------------------------------------------------",[122,160,0],"#------------------------------------------------------------------------------------------------",[115,161,0],"#------------------------------------------------------------------------------------------------",[116,161,0],"#------------------------------------------------------------------------------------------------",[117,161,0],"#------------------------------------------------------------------------------------------------",[118,161,0],"#------------------------------------------------------------------------------------------------",[117,162,0],"#------------------------------------------------------------------------------------------------",[118,162,0],"#------------------------------------------------------------------------------------------------",[117,163,0],"#------------------------------------------------------------------------------------------------",[118,163,0],"#------------------------------------------------------------------------------------------------",[117,164,0],"#------------------------------------------------------------------------------------------------",[118,164,0],"#------------------------------------------------------------------------------------------------",[119,164,0],"#------------------------------------------------------------------------------------------------",[120,164,0],"#------------------------------------------------------------------------------------------------",[121,164,0],"#------------------------------------------------------------------------------------------------",[122,164,0],"#------------------------------------------------------------------------------------------------"]
```

kevingranade committed:
Remove the hordes option

kevingranade committed:
Shift hordes visualization features to new data structures

kevingranade committed:
Extract horde data structures to a dedicated class that partitions monsters by behavior.

i.e. active monsters go to a seperate container that is the only thing referenced by move_hordes()

kevingranade commented:
While I was looking into implementing priority queue stuff, I realized that we can avoid walking the idle horde data by partitioning entities across multiple (as it turns out, three) containers.  Due this adding complexity to interactions, I decided it was now mandatory to take this now more complex data and encapsulate it, so now we have class horde_map;

The per-turn overhead is now 0 in the very common case that you haven't alerted any horde entities, this is subject to change in the future, but that's also going to call for the scheduler stuff to be added.

This class abstracts over there being three different containers, by default if you iterate over the container (via begin() end() ) you get everything, but if you get a lightweight "view" object it can filter it to any subset of active, idle, and dormant horde entities.  So for move_hordes() it only walks the active hordes, for seeing hordes and obstructing movement it only walks active and idle hordes, and for signal_hordes and spawning horde entities onto the map, it walks everything.
One kind of clunky thing this causes is that if you want to set an idle entity to be active or make an active entity idle, you have to remove and re-insert the node into the new container.  The good thing is this doesn't do any meaningful allocation since it can splice the nodes between maps.

kevingranade committed:
Add test for dormant zombie interaction with hordes

kevingranade committed:
Add clear overmaps helper

kevingranade committed:
Bulk fixup for IWYU

Brambor commented on src/horde_map.h:
```diff
@@ -0,0 +1,257 @@
+                    if( !( filter & horde_map_flavors::active ) || outer_map->empty() ) {
+                        do {
+                            next_map();
+                            if( outer_map == nullptr ) {
+                                return;
+                            }
+                        } while( outer_map->empty() );
```
```suggestion
                        } while( outer_map != nullptr && outer_map->empty() );outer_map->empty() );
```

kevingranade replied:
This actually does need to be separate because if outer_map equals nullptr here it means the whole container is empty and we want to exit, but if outer_map->empty() == true it just means the current sub-map is empty and a later one might not be, so we try again.

Brambor replied:
I believe my change is a refactor:
1. `outer_map == nullptr`:
   - `while(outer_map != nullptr && outer_map->empty())` :=
   - `while(nullptr != nullptr && outer_map->empty())` :=
   - `while(false && outer_map->empty())` :=
   - `while(false)`
   - ends the cycle
1. `outer_map.empty() is true`:
   - `while(outer_map != nullptr && outer_map->empty())` :=
   - `while(true && true)` :=
   - next cycle
1. `outer_map.empty() is false`:
   - `while(outer_map != nullptr && outer_map->empty())` :=
   - `while(true && false)` :=
   - ends the cycle

Unless I am blundering. The same is true for the original code.

Brambor replied:
I see my blunder now!

db48x commented on src/sdltiles.cpp:
```diff
@@ -896,40 +895,32 @@ void cata_tiles::draw_om( const point &dest, const tripoint_abs_omt &center_abs_
                         if( find_tile_with_season( id ) ) {
                             // NOLINTNEXTLINE(cata-translate-string-literal)
-                            draw_from_id_string( string_format( "overmap_horde_%d", horde_size < 10 ? horde_size : 10 ),
+                            draw_from_id_string( string_format( "overmap_horde_%d", horde_size < 32 ? horde_size : 10 ),
```
That makes sense. Perhaps the comment could say that instead of calling it a fallback. :)

If you end up with another switch statement, can you extract the switches into separate functions? Simplifying the code would make it much more understandable:

```c++
if( find_tile_with_season( id ) ) {
    // If there's a "seasonal variant" (horde specific) sprite set map to that.
    sprite = get_horde_sprite_name( horde_size );
} else {
    // Else use the sprite of the default monster in the horde group, zombies.
    sprite = get_zombie_sprite_name( horde_size );
}
draw_from_id_string( sprite, omp, 0, 0, lit_level::LIT, false );
```

Almost don’t even need the comments when it’s that simple.

kevingranade replied:
Yea that's a good idea. I think applying it in two stages is good:
int one_to_ten = scale_horde_population_to_sprite_size( population )
sprite = get_*_sprite_name( one_to_ten );

db48x replied:
Ooh, that’s actually nicer. But I notice that in your latest revision you haven’t yet put everything into separate functions.

kevingranade replied:
Yea I was looking at making it universal between tiles and curses then got sidetracked wrapping up clang tidy and the last build issues.

Extending the curses visualization is a can of worms I don't want to open so I'll loop around and at least extract it to functions the next time I'm at my computer.

kevingranade commented:
Test build available at https://github.com/CleverRaven/Cataclysm-DDA/releases/tag/cdda-experimental-2025-09-20-1932

kevingranade committed:
Tweak horde display logic

kevingranade committed:
tidy fix, overmap_map_data_cache

kevingranade committed:
tidy fix, overmap

ehughsbaird commented on src/init.cpp:
```diff
@@ -405,6 +406,7 @@ void DynamicDataLoader::initialize()
     add( "overmap_special_migration", &overmap_special_migration::load_migrations );
     add( "city_building", &city_buildings::load );
     add( "map_extra", &MapExtras::load );
+    add( "omt_placeholder", &map_data_placeholders::load );
```
Please also add a finalize function, which calls finalize on the factory. This ensures that deferred JSON is cleared.

kevingranade replied:
Done

ehughsbaird commented on src/map.cpp:
```diff
@@ -8057,6 +8066,28 @@ void map::shift( const point_rel_sm &sp )
+    // TODO: can probably skip a bunch of iteration here.
+    for( int gridx = x_start; gridx != x_stop; gridx += x_step ) {
+        for( int gridy = y_start; gridy != y_stop; gridy += y_step ) {
+            if( ( sp.x() && gridx == x_start ) || ( sp.y() && gridy == y_start ) ) {
```
Why the x != 0 and y != 0 condition?

ehughsbaird commented on src/overmap.cpp:
```diff
@@ -946,6 +951,39 @@ void oter_type_t::load( const JsonObject &jo, const std::string_view )
 
+    if( jo.has_string( "default_map_data" ) ) {
+        optional( jo, was_loaded, "default_map_data", default_map_data );
+    } else {
```
This prevents copy-from from working with this, because it will always be overwritten.
```suggestion
    } else if( !was_loaded ) {
```

kevingranade replied:
To clarify, if it's not a copy-from entry, was_loaded will be false at this point and apply the fallback, if it IS a copy-from entry it will be true and only overwrite if there is a "default_map_data" field?
(I'm seeing a seemingly related crash after applying this)

ehughsbaird replied:
Yes. To ensure that the optional isn't default initializing default_map_data (for some reason), consider replacing it with `mandatory` (because it is when the member is defined).

example:
B copies-from A.

Loading A creates a new entity, so `A.was_loaded` is false.
A defines `"default_map_data"` `foo`.
after loading, `A.was_loaded` is now true.

Loading B creates an entity that is a copy of A, so `B.was_loaded` is true.
If B does not define `"default_map_data"`, this else case is hit, and it does not keep `foo`, as it should.

ehughsbaird commented on src/overmap.cpp:
```diff
@@ -946,6 +951,39 @@ void oter_type_t::load( const JsonObject &jo, const std::string_view )
     optional( jo, was_loaded, "connect_group", connect_group, string_reader{} );
     optional( jo, was_loaded, "travel_cost_type", travel_cost_type, oter_travel_cost_type::other );
 
+    if( jo.has_string( "default_map_data" ) ) {
```
I prefer these checks to be `has_member`, and let `optional` complain if it's not the format it likes, instead of silently ignoring it.
```suggestion
    if( jo.has_member( "default_map_data" ) ) {
```

ehughsbaird commented on src/overmap.cpp:
```diff
@@ -4328,7 +4488,7 @@ void overmap::signal_hordes( const tripoint_rel_sm &p_rel, const int sig_power )
             continue;
         }
-        const int dist = rl_dist( absp, mg.abs_pos );
+        const int dist = 12 * rl_dist( absp, mg.abs_pos );
```
Should this 12 be SEEX?

kevingranade replied:
Thanks for bringing this up, it was actually broken :D
Changed the 12s to SEEX and commented a bit why,

ehughsbaird commented on src/overmap.cpp:
```diff
@@ -4360,6 +4520,46 @@ void overmap::signal_hordes( const tripoint_rel_sm &p_rel, const int sig_power )
+        tripoint_abs_ms origin = project_to<coords::ms>( absp ) + point{ 6, 6 };
+        bool already_active = mon->second.is_active();
+        const int dist = rl_dist( origin, mon->first );
+        if( sig_power * 12 < dist ) {
```
Ditto, and below?

kevingranade commented:
New blocker yay!
Forwarding sound events to hordes is horrifically non-performant now.  The first step is filtering for distance at the OMT level, after that comes deferring delivery with some kind of queuing system and the scheduler I mentioned before, I hope it doesn't come to that.

kevingranade committed:
Overhaul sound delivery to hordes to allow filtering out distant hordes cheaply.

kevingranade commented:
I did the straightforward option of simply filtering out submap sized chunks of hordes if they are below the trigger threshold and now sound delivery isn't even showing up in profiling.

This code was actually riddled with bugs so this turned out to be a very good thing, in the process I cleaned up the code path delivering sounds to hordes and removed several points that were only succeeding in specific circumstances.

kevingranade commented on data/json/monstergroups/zombies.json:
```diff
@@ -91,6 +91,58 @@
   },
+  {
+    "type": "monstergroup",
+    "id": "GROUP_ZOMBIE_HORDE",
```
Once we have different behaviors for different horde entities we can re-add panicked persons to these groups or go back to using the existing group, but that's too much scope creep for this PR.

kevingranade committed:
Rerange some hordes tests to avoid expensive overmap wipes

GuardianDll commented:
I trust everyone's comment here pointed to very important matters, but the PR is way too big already, so please adress the problems in new issues and fixes in consequent PRs, ty all
I tested PR a bit, i perfed it, it looks good in my eyes, so good job
